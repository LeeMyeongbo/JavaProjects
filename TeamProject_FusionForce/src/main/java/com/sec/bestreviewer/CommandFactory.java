package com.sec.bestreviewer;

import com.sec.bestreviewer.command.*;
import com.sec.bestreviewer.store.Employee;
import com.sec.bestreviewer.store.EmployeeStore;
import com.sec.bestreviewer.util.OptionParser;
import com.sec.bestreviewer.util.Pair;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CommandFactory {

    public static final String CMD_ADD = "ADD";
    public static final String CMD_DEL = "DEL";
    public static final String CMD_SCH = "SCH";
    public static final String CMD_CNT = "CNT";
    public static final String CMD_MOD = "MOD";

    private static final String EMPLOYEE_NUMBER = "employeeNum";
    private static final String NAME = "name";
    private static final String CAREER_LEVEL = "cl";
    private static final String PHONE_NUMBER = "phoneNum";
    private static final String BIRTHDAY = "birthday";
    private static final String CERTI = "certi";

    private static final Map<String, String> fieldMap = new HashMap<>();

    static {
        fieldMap.put(EMPLOYEE_NUMBER, EmployeeStore.FIELD_EMPLOYEE_NUMBER);
        fieldMap.put(NAME, EmployeeStore.FIELD_NAME);
        fieldMap.put(CAREER_LEVEL, EmployeeStore.FIELD_CAREER_LEVEL);
        fieldMap.put(PHONE_NUMBER, EmployeeStore.FIELD_PHONE_NUMBER);
        fieldMap.put(BIRTHDAY, EmployeeStore.FIELD_BIRTH_DAY);
        fieldMap.put(CERTI, EmployeeStore.FIELD_CERTI);
    }

    private CommandFactory() {
    }

    public static Command buildCommand(String cmd, List<String> options, List<String> params)
            throws IllegalArgumentException {
        final OptionParser optionParser = new OptionParser(options);

        return switch (cmd) {
            case CMD_ADD -> {
                final Employee employee = new Employee(params.get(0), params.get(1), params.get(2),
                    params.get(3), params.get(4), params.get(5));
                yield new AddCommand(optionParser, employee);
            }
            case CMD_DEL -> new DeleteCommand(optionParser, getConditionMapFromParams(params));
            case CMD_SCH -> new SearchCommand(optionParser, getConditionMapFromParams(params));
            case CMD_CNT -> new CountCommand();
            case CMD_MOD -> new ModifyCommand(optionParser, getConditionMapFromParams(params));
            default -> throw new IllegalArgumentException("Wrong command");
        };
    }

    private static List<Pair<String, String>> getConditionMapFromParams(List<String> params) {
        List<Pair<String, String>> conditionPairs = new ArrayList<>();
        for (int i = 0; i < params.size(); i += 2) {
            final String fieldName = fieldMap.get(params.get(i));
            if (fieldName == null) {
                throw new IllegalArgumentException("Wrong field: " + params.get(i));
            }
            conditionPairs.add(Pair.create(fieldName, params.get(i + 1)));
        }
        return conditionPairs;
     }
}
