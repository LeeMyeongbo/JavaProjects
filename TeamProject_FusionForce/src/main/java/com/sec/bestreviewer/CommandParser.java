package com.sec.bestreviewer;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CommandParser {

    private final static int MIN_TOKENS_NUM = 5;
    private final static String SPACE = " ";
    private final static String EMPTY = "";

    public TokenGroup parse(String line) throws IllegalArgumentException {
        String[] r = line.split(",", -1);

        if (r.length < MIN_TOKENS_NUM) {
            throw new IllegalArgumentException("wrong command format");
        }

        String type = r[0];
        List<String> options = new ArrayList<>();
        List<String> params = new ArrayList<>();

        if (r.length > 8 && (r[6].equals("-a") || r[6].equals("-o"))) {
            options.addAll(Arrays.asList(Arrays.copyOfRange(r, 1, 4)));
            options.addAll(Arrays.asList(Arrays.copyOfRange(r, 6, 9)));
            params.addAll(Arrays.asList(Arrays.copyOfRange(r, 4, 6)));
            params.addAll(Arrays.asList(Arrays.copyOfRange(r, 9, r.length)));
        } else {
            options.addAll(Arrays.asList(Arrays.copyOfRange(r, 1, 4)));
            params.addAll(Arrays.asList(Arrays.copyOfRange(r, 4, r.length)));
        }

        return new TokenGroup(type, options, getValidList(params));
    }

    private List<String> getValidList(List<String> list) {
        List<String> validList = new ArrayList<>();

        list.forEach(i -> {
            if (!EMPTY.equals(i) && !SPACE.equals(i)) {
                validList.add(i);
            }
        });

        return validList;
    }
}
