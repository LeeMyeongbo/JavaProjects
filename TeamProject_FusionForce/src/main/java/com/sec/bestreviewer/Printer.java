package com.sec.bestreviewer;

import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

public class Printer {

    private final String outputFilName;
    private final StringBuilder stdStringBuilder;

    public Printer(String outputFilename) {
        this(outputFilename, new StringBuilder());
    }

    public Printer(String outputFilename, StringBuilder stdStringBuilder) {
        this.outputFilName = outputFilename;
        this.stdStringBuilder = stdStringBuilder;
    }

    public void add(List<String> lines) {
        for (String line : lines) {
            stdStringBuilder.append(line).append("\n");
        }
    }

    public void printOutputFile() {
        try (OutputStream output = Files.newOutputStream(Paths.get(outputFilName))) {
            output.write(stdStringBuilder.toString().getBytes());
        } catch (Exception e) {
            e.getStackTrace();
        }
    }
}
