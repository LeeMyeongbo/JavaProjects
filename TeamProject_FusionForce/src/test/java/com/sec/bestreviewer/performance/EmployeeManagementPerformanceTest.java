package com.sec.bestreviewer.performance;

import com.sec.bestreviewer.EmployeeManagement;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.text.DecimalFormat;

public class EmployeeManagementPerformanceTest {

    @Test
    public void deletePerformanceTest() {
        final String outputFileName = "./src/test/java/com/sec/bestreviewer/performance/delete_performance_test_output.txt";
        final String inputFileName = "./src/test/java/com/sec/bestreviewer/performance/delete_performance_test_input.txt";

        String[] args = {inputFileName, outputFileName};
        File outputFile = new File(outputFileName);

        EmployeeManagement employeeManagement = new EmployeeManagement();

        long startTime = System.nanoTime();
        employeeManagement.run(args);
        long endTime = System.nanoTime();

        double durationSec = (double)(endTime - startTime) / 1000000000.0; // 나노초를 초로 변환
        DecimalFormat df = new DecimalFormat("#.###"); // 소수점 세 자리까지 출력
        System.out.println("Execution time: " + df.format(durationSec) + "s");

        Approvals.verify(outputFile);

    }

    @Test
    public void searchPerformanceTest() {
        final String outputFileName = "./src/test/java/com/sec/bestreviewer/performance/search_performance_test_output.txt";
        final String inputFileName = "./src/test/java/com/sec/bestreviewer/performance/search_performance_test_input.txt";

        String[] args = {inputFileName, outputFileName};
        File outputFile = new File(outputFileName);

        EmployeeManagement employeeManagement = new EmployeeManagement();

        long startTime = System.nanoTime();
        employeeManagement.run(args);
        long endTime = System.nanoTime();

        double durationSec = (double)(endTime - startTime) / 1000000000.0; // 나노초를 초로 변환
        DecimalFormat df = new DecimalFormat("#.###"); // 소수점 세 자리까지 출력
        System.out.println("Execution time: " + df.format(durationSec) + "s");

        Approvals.verify(outputFile);
    }

    @Test
    public void searchThirdOptionPerformanceTest() {
        final String outputFileName = "./src/test/java/com/sec/bestreviewer/performance/search_third_option_performance_test_output.txt";
        final String inputFileName = "./src/test/java/com/sec/bestreviewer/performance/search_third_option_performance_test_input.txt";

        String[] args = {inputFileName, outputFileName};
        File outputFile = new File(outputFileName);

        EmployeeManagement employeeManagement = new EmployeeManagement();

        long startTime = System.nanoTime();
        employeeManagement.run(args);
        long endTime = System.nanoTime();

        double durationSec = (double)(endTime - startTime) / 1000000000.0; // 나노초를 초로 변환
        DecimalFormat df = new DecimalFormat("#.###"); // 소수점 세 자리까지 출력
        System.out.println("Execution time: " + df.format(durationSec) + "s");

        Approvals.verify(outputFile);
    }

    @Test
    public void modifyPerformanceTest() {
        final String outputFileName = "./src/test/java/com/sec/bestreviewer/performance/modify_performance_test_output.txt";
        final String inputFileName = "./src/test/java/com/sec/bestreviewer/performance/modify_performance_test_input.txt";

        String[] args = {inputFileName, outputFileName};
        File outputFile = new File(outputFileName);

        EmployeeManagement employeeManagement = new EmployeeManagement();

        long startTime = System.nanoTime();
        employeeManagement.run(args);
        long endTime = System.nanoTime();

        double durationSec = (double)(endTime - startTime) / 1000000000.0; // 나노초를 초로 변환
        DecimalFormat df = new DecimalFormat("#.###"); // 소수점 세 자리까지 출력
        System.out.println("Execution time: " + df.format(durationSec) + "s");


        Approvals.verify(outputFile);
    }
}
