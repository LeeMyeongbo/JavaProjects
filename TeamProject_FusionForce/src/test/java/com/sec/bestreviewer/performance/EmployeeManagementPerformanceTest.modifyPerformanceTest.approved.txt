MOD,90006103,XSZZXK KIM,CL1,010-5899-3939,20010307,ADV
MOD,92000576,RTGTUZ KIM,CL4,010-2956-0057,20040721,PRO
MOD,92008675,JXSVWJ KIM,CL4,010-1075-8883,19841224,ADV
MOD,95003063,MTFWZW FTO,CL3,010-0150-7385,19860921,EX
MOD,04009930,DLVIVP KIM,CL3,010-3329-6264,19780905,PRO
MOD,95001593,VBVFEC ENM,CL4,010-2116-6687,20040127,PRO
MOD,95004504,NZSAAF PFV,CL3,010-2116-9366,19770405,PRO
MOD,01005182,JALFTA IHU,CL1,010-2116-0082,20000219,ADV
MOD,04001821,RSYGFH GPD,CL2,010-2116-8057,20010727,PRO
MOD,08007958,UNRRIU SLJ,CL4,010-2116-6023,19980207,EX
SCH,10
MOD,92001857,ILYCOG BEC,CL2,010-6332-3761,20151020,ADV
MOD,92005800,WHRUXG MFO,CL1,010-9067-3761,20080714,PRO
MOD,92006551,OSHCCM ENN,CL3,010-7160-3761,19780309,EX
MOD,92008470,ARQGAP HKX,CL1,010-7770-3761,19750520,PRO
MOD,95004472,WVPEOV DUL,CL3,010-3605-3761,19960219,PRO
SCH,20
MOD,90000053,PRCBZH XXE,CL2,010-2746-9204,19910909,ADV
MOD,90000711,BJPYCK MDA,CL2,010-1743-1341,19910421,ADV
MOD,90001329,ZCXHYT IWR,CL2,010-3518-2421,19911008,EX
MOD,90001429,GPFGQF PSN,CL2,010-0029-6234,19911016,ADV
MOD,90001508,IAWPLA NBN,CL2,010-8559-7943,19910328,EX
MOD,90000026,UKQPIV MGF,CL1,010-1475-4996,19750515,PRO
MOD,90000076,FDSPXT TJG,CL1,010-9918-9933,20140504,EX
MOD,90000090,RDVQHY JJW,CL1,010-9546-5033,20180521,PRO
MOD,90000166,XACYNM NJB,CL1,010-8908-2413,19820506,PRO
MOD,90000223,NLAJHM LPE,CL1,010-4896-7591,20050502,PRO
MOD,90000132,NWRATR TXG,CL3,010-0091-0607,19921015,PRO
MOD,90000133,RBYGFU QZI,CL3,010-3925-2552,20190615,PRO
MOD,90000177,EDFDOA NCV,CL3,010-8949-4804,20070415,ADV
MOD,90000464,GXPPCA NRS,CL3,010-7024-5207,20190715,ADV
MOD,90000490,EOLKXP CTT,CL3,010-7489-0523,19950515,PRO
SCH,33126
MOD,33978
SCH,67104
